src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Ricci.cpp

```diff
@@ -0,0 +1,163 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Ricci.hpp"
+
+#include "DataStructures/DataVector.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/TempBuffer.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+
+/// \cond
+namespace GeneralizedHarmonic {
+template <size_t VolumeDim, typename Frame, typename DataType>
+void spatial_ricci_tensor(
+    const gsl::not_null<tnsr::ii<DataType, VolumeDim, Frame>*> ricci,
+    const tnsr::iaa<DataType, VolumeDim, Frame>& phi,
+    const tnsr::ijaa<DataType, VolumeDim, Frame>& deriv_phi,
+    const tnsr::II<DataType, VolumeDim, Frame>&
+        inverse_spatial_metric) noexcept {
+  if (UNLIKELY(get_size(get<0, 0>(*ricci)) != get_size(get<0, 0, 0>(phi)))) {
+    *ricci = tnsr::ii<DataType, VolumeDim, Frame>(get_size(get<0, 0, 0>(phi)));
+  }
+
+  TempBuffer<tmpl::list<::Tags::TempIjj<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempijk<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempi<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempi<1, VolumeDim, Frame, DataType>,
+                        ::Tags::TempI<0, VolumeDim, Frame, DataType>,
+                        ::Tags::TempI<1, VolumeDim, Frame, DataType>>>
+      local_buffer(get_size(get<0, 0>(inverse_spatial_metric)));
+
+  // New variable to avoid recomputing in nested loops
+  auto& phi_uplolo =
+      get<::Tags::TempIjj<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = i; j < VolumeDim; ++j) {  // Symmetry
+        phi_uplolo.get(k, i, j) = 0.;
```

Comment:
`phi_uplolo.get(k, i, j) =  0.5 * inverse_spatial_metric.get(k, 0) * phi.get(0, 1 + i, 1 + j);`, then loop starting at 1. This saves an assignment of the full data structure.

Reply:
changed, thanks.